utils/zerovec/src/ule/niche.rs

```diff
@@ -0,0 +1,187 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core::marker::Copy;
+use core::mem::size_of;
+
+use super::{AsULE, ULE};
+
+/// The [`ULE`] types implementing this trait guarantee that [`NicheBytes::NICHE_BIT_PATTERN`]
+/// can never occur as a valid byte representation of the type.
+///
+/// Guarantees for a valid implementation.
+/// 1. N must be equal to core::mem::sizeo_of::<Self>() or else it will
+///    cause panics.
+/// 2. The bit pattern [`NicheBytes::NICHE_BIT_PATTERN`] must not be incorrect as it would lead to
+///    weird behaviour.
+/// 3. The abstractions built on top of this trait must panic on an invalid N.
+/// 4. The abstractions built on this trait that use type punning must ensure that type being
+///    punned is [`ULE`].
+pub trait NicheBytes<const N: usize> {
+    const NICHE_BIT_PATTERN: [u8; N];
+}
+
+/// [`ULE`] type for [`NichedOption<U,N>`] where U implements [`NicheBytes`].
+/// The invalid bit pattern is used as the niche.
+///
+/// This uses 1 byte less than [`crate::ule::OptionULE<U>`] to represent [`NichedOption<U,N>`]
+/// which Derefs to [`Option<U>`].
+/// The implementors guarantee that N == core::mem::sizeo_of::<Self>()
+///
+/// # Example
+///
+/// ```
+/// use zerovec::ZeroVec;
+/// use zerovec::ule::NichedOption;
+/// use core::num::NonZeroI8;
+///
+/// let bytes = &[0x00, 0x01, 0x02, 0x00 ];
+/// let zv_no: ZeroVec<NichedOption<NonZeroI8, 1>> =
+///         ZeroVec::parse_byte_slice(bytes).expect("Unable to parse as NichedOption.");
+///
+/// assert_eq!(zv_no.get(0).map(|e| e.0), Some(None));
+/// assert_eq!(zv_no.get(1).map(|e| e.0), Some(NonZeroI8::new(1)));
+/// assert_eq!(zv_no.get(2).map(|e| e.0), Some(NonZeroI8::new(2)));
+/// assert_eq!(zv_no.get(3).map(|e| e.0), Some(None));
+/// ```
+// Invariants:
+// The union stores [`NicheBytes::NICHE_BIT_PATTERN`] when None.
+// Any other bit pattern is a valid.
+#[repr(C)]
+pub union NichedOptionULE<U: NicheBytes<N> + ULE, const N: usize> {
+    /// Invariant: The value is `niche` only if the bytes equal NICHE_BIT_PATTERN.
+    niche: [u8; N],
+    /// Invariant: The value is `valid` if the `niche` field does not match NICHE_BIT_PATTERN.
+    valid: U,
+}
+
+impl<U: NicheBytes<N> + ULE, const N: usize> NichedOptionULE<U, N> {
+    /// New NichedOptionULE<U, N> from Option<U>
+    pub fn new(opt: Option<U>) -> Self {
+        assert!(N == core::mem::size_of::<U>());
```

Comment:
Can you make this a const assert?

Reply:
I'm not sure if we can due to `error[E0401]: can't use generic parameters from outer function`